src/ctia/shutdown.clj

```diff
@@ -0,0 +1,22 @@
+(ns ctia.shutdown
+  (:require [clojure.tools.logging :as log]))
+
+(defonce shutdown-hooks (atom {:registered? false :hooks {}}))
+
+(defn register-hook! [key f]
+  (swap! shutdown-hooks assoc-in [:hooks key] f))
+
+(defn shutdown-ctia! []
+  (doseq [[name hook] (:hooks @shutdown-hooks)]
+    (try
+      (hook)
+      (catch Exception excep
+        (log/error excep (format "Shutdown hook %s failed" name)))))
+  (swap! shutdown-hooks assoc :hooks {}))
+
+(defn init! []
+  (when-not (:registered? @shutdown-hooks)
+    (.addShutdownHook
+     (Runtime/getRuntime)
+     (Thread. shutdown-ctia!))
+    (swap! shutdown-hooks assoc :registered? true)))
```

Comment:
I think there could still be a race condition here if we call `init!` concurrently. I'm not sure this is really necessary as `init!` shouldn't be called concurrently. But just in case I believe we could prevent any problem by using something like:

~~~clojure
(let [can-add-hook (ref false)]
       (dosync
           (when-not (:registered? @shutdown-hooks)
                (swap! shutdown-hook assoc registered? true)
                (ref-set can-add-hook true)))
        (when @can-add-hook 
            (.addShutdownHook
                 (Runtime/getRuntime)
                 (Thread. shutdown-ctia)))
~~~